### PR TITLE
[feature-wip](multi-catalog)(step2)support read max compute data by JNI

### PR DESCRIPTION
**Pull request by wsjz**
# Proposed changes

 Issue Number: #19679

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/exec/scan/vfile_scanner.cpp [hunk 1]
```diff
@@ -577,6 +578,15 @@ Status VFileScanner::_get_next_reader() {
         Status init_status;
         // TODO: use data lake type
         switch (_params.format_type) {
+        case TFileFormatType::FORMAT_JNI: {
```
Comment (AshinGau) [inline be/src/vec/exec/scan/vfile_scanner.cpp:581]: Two layer nested logic:
```
if (TFileFormatType::FORMAT_JNI) {
  // do jni scanner
  if (range.table_format_params.table_format_type == "maxcompute") {
    // max compute
  } else if (range.table_format_params.table_format_type == "hudi") {
    // hudi
  }
}
```

File: fe/java-udf/src/main/java/org/apache/doris/jni/vec/ScanPredicate.java [hunk 2]
```diff
@@ -119,6 +119,11 @@ public String toString() {
             return inspectObject().toString();
         }
 
+        @Override
+        public boolean isNull() {
+            return valueBytes == null || valueBytes.length == 0;
```
Comment (AshinGau) [inline fe/java-udf/src/main/java/org/apache/doris/jni/vec/ScanPredicate.java:124]: return false always

**Comment by wsjz** [comment]
run buildall

**Review by AshinGau — DISMISSED** [review]

**Comment by AshinGau** [comment]
LGTM

**Comment by wsjz** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/FileQueryScanNode.java [hunk 3]
```diff
@@ -193,7 +193,10 @@ private void setColumnPositionMapping()
             throws UserException {
         TableIf tbl = getTargetTable();
         List<Integer> columnIdxs = Lists.newArrayList();
-
+        if (params.getRequiredSlots() == null) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/FileQueryScanNode.java:196]: What is this for？ Add some comment to explain it.

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/MaxComputeScanNode.java [hunk 4]
```diff
@@ -74,7 +78,37 @@ protected Map<String, String> getLocationProperties() throws UserException {
...
-        result.add(new FileSplit(new Path("/"), 0, -1, -1, 0L, new String[0], Collections.emptyList()));
+        // String splitPath = catalog.getTunnelUrl();
+        // TODO: use single max compute scan node rather than file scan node
+        com.aliyun.odps.Table odpsTable = table.getOdpsTable();
+        if (desc.getSlots().isEmpty() || odpsTable.getFileNum() <= 0) {
+            return result;
+        }
+        try {
+            List<Pair<Long, Long>> sliceRange = new ArrayList<>();
+            long totalRows = catalog.getTotalRows(table.getDbName(), table.getName());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/MaxComputeScanNode.java:89]: Is this `getTotalRows()` method expensive?

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/MaxComputeScanNode.java [hunk 5]
```diff
@@ -74,7 +78,37 @@ protected Map<String, String> getLocationProperties() throws UserException {
...
+            long fileNum = odpsTable.getFileNum();
+            long splitSize = (long) Math.ceil((double) totalRows / fileNum);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/MaxComputeScanNode.java:91]: If the `totalRows` is less than `fileNum`, the `splitSize` will be 0.

Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/planner/external/MaxComputeScanNode.java:91]: And what if there is only 100 rows, and 10 files. Do we need to split into 10 ranges? I think this is too small.

File: fe/java-udf/src/main/java/org/apache/doris/jni/vec/VectorColumn.java [hunk 6]
```diff
@@ -21,6 +21,8 @@
 import org.apache.doris.jni.utils.TypeNativeBytes;
 import org.apache.doris.jni.vec.ColumnType.Type;
 
+import org.apache.log4j.Logger;
```
Comment (morningman) [inline fe/java-udf/src/main/java/org/apache/doris/jni/vec/VectorColumn.java:24]: Wrong logger

File: be/src/vec/exec/jni_connector.cpp [hunk 7]
```diff
@@ -63,14 +63,15 @@ JniConnector::~JniConnector() {
...
+    if (env == nullptr) {
         return Status::InternalError("Failed to get/create JVM");
     }
-    RETURN_IF_ERROR(_init_jni_scanner(_env, state->batch_size()));
+    RETURN_IF_ERROR(_init_jni_scanner(env, state->batch_size()));
     // Call org.apache.doris.jni.JniScanner#open
-    _env->CallVoidMethod(_jni_scanner_obj, _jni_scanner_open);
-    RETURN_ERROR_IF_EXC(_env);
+    env->CallVoidMethod(_jni_scanner_obj, _jni_scanner_open);
+    RETURN_ERROR_IF_EXC(env);
```
[the remark this answers is not in the document]
Comment (wsjz) [reply be/src/vec/exec/jni_connector.cpp:74]: env is thread local in jvm

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/MaxComputeScanNode.java (hunk 4 above)
Comment (wsjz) [reply fe/fe-core/src/main/java/org/apache/doris/planner/external/MaxComputeScanNode.java:89]: access metaservice， just call when catalog init

**Comment by wsjz** [comment]
run buildall

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**